src/main/java/org/folio/rest/impl/UsersAPI.java

```diff
@@ -228,40 +238,47 @@ public void postUsers(String lang, User entity, RoutingContext routingContext,
     entity.setUpdatedDate(now);
-    PgUtil.post(TABLE_NAME_USERS, entity, okapiHeaders, vertxContext, PostUsersResponse.class, reply -> {
-      if (isDuplicateIdError(reply)) {
-        asyncResultHandler.handle(
-          succeededFuture(PostUsersResponse.respond422WithApplicationJson(
-            ValidationHelper.createValidationErrorMessage(
-              "id", entity.getId(),
-              "User with this id already exists"))));
-        return;
-      }
-      if (isDuplicateUsernameError(reply)) {
-        asyncResultHandler.handle(
-          succeededFuture(PostUsersResponse.respond422WithApplicationJson(
-            ValidationHelper.createValidationErrorMessage(
-              "username", entity.getUsername(),
-              "User with this username already exists"))));
-        return;
-      }
-      if (isDuplicateBarcodeError(reply)) {
-        asyncResultHandler.handle(
-          succeededFuture(PostUsersResponse.respond422WithApplicationJson(
-            ValidationHelper.createValidationErrorMessage(
-              "barcode", entity.getBarcode(),
-              "This barcode has already been taken"))));
-        return;
-      }
-      logger.debug("Save successful");
-      asyncResultHandler.handle(reply);
-    });
+    String userId = StringUtils.defaultIfBlank(entity.getId(), UUID.randomUUID().toString());
+
+    pgClient.withTrans(conn -> conn.saveAndReturnUpdatedEntity(TABLE_NAME_USERS, userId, entity.withId(userId))
+      .compose(user -> userOutboxService.saveUserOutboxLog(conn, user, UserEvent.Action.CREATE, okapiHeaders)
+        .map(aVoid -> PostUsersResponse.respond201WithApplicationJson(entity, PostUsersResponse.headersFor201().withLocation(userId)))
+        .map(Response.class::cast))
+      .onComplete(reply -> {
+        if (isDuplicateIdError(reply)) {
+          asyncResultHandler.handle(
+            succeededFuture(PostUsersResponse.respond422WithApplicationJson(
+              ValidationHelper.createValidationErrorMessage(
+                "id", entity.getId(),
+                "User with this id already exists"))));
+          return;
+        }
+        if (isDuplicateUsernameError(reply)) {
+          asyncResultHandler.handle(
+            succeededFuture(PostUsersResponse.respond422WithApplicationJson(
+              ValidationHelper.createValidationErrorMessage(
+                "username", entity.getUsername(),
+                "User with this username already exists"))));
+          return;
+        }
+        if (isDuplicateBarcodeError(reply)) {
+          asyncResultHandler.handle(
+            succeededFuture(PostUsersResponse.respond422WithApplicationJson(
+              ValidationHelper.createValidationErrorMessage(
+                "barcode", entity.getBarcode(),
+                "This barcode has already been taken"))));
+          return;
+        }
+        logger.debug("Save successful");
+        userOutboxService.processOutboxEventLogs(vertxContext.owner(), okapiHeaders);
```

Comment:
My interpretation of this code is that it processes the outbox at the end of handling the API request to create a user.

It is my understanding that the intention of the outbox pattern is that messages are sent separately in a durable process. This seems counter to that pattern. What is the reason for trigger the processing of the outbox during each invocation of the create user API?

Reply:
This is done to send changes to Kafka immediately after saving. If an error occurs during sending, the processOutboxEvent will be called from the API request handler.

Reply:
I can understand how it's useful for that to be done eagerly in order to reduce potential lag for publishing the message. 

I think that somewhat contradicts how the outbox pattern is intended to work. In this case, it presents a challenge with knowing that the background processing is working. 

Given that the message is published during the API request, how do folks know that the background outbox publishing process is working?